sources/csv_reader/__init__.py

```diff
@@ -0,0 +1,70 @@
+"""
+Source, which uses the `filesystem` source and DuckDB
+to extract CSV files data from the given locations.
+"""
+from typing import Iterable, List
+
+import duckdb
+import fsspec
+import pendulum
+
+import dlt
+from dlt.common.typing import TDataItem, TAnyDateTime
+
+try:
+    from filesystem import filesystem
+except ImportError:
+    from sources.filesystem import filesystem
+
+from fsspec.implementations.local import LocalFileSystem
+
+from .helpers import add_columns
+
+
+@dlt.resource
+def read_location(files):
+    """A resource to extract data from the given CSV files.
+
+    Args:
+        files (List[FileItem]): A list of files to read.
+
+    Returns:
+        Iterable[TDataItem]: Data items, read from the given CSV files.
+    """
+    state = dlt.current.resource_state()
+    start_from = state.setdefault("last_modified", pendulum.datetime(1970, 1, 1))
+
+    results = []
+    for file in files:
+        if file["modification_date"] <= start_from:
+            continue
+
+        file_res = duckdb.sql(f"""SELECT * FROM read_csv_auto('{file["file_url"]}')""")
```

Comment:
please try `duckdb.read_csv()`. what I expect to happen there
1. you can possibly pass opened file as name parameter. we have a chance that any fsspec open file will work without registering filesystems in duck db
2. the method returns duckb relation object which you can read in batches. your code below will load all records in one go, this is not suitable for production

in other words: 
1. let's get reading and yielding batches straight
2. we should be able to yield arrow tables or arrow batches - this will also carry the schema to dlt and I think it is our preferred method
3. let's make sure we can use fsspec opened objects instead of providing names

if (3) is true we'll move this code to `filesystem` verified source (or merge the filesystem to the core library...) but let's do another iteration here

Reply:
@rudolfix, I don't think the item 3 is that easy. Those objects returned from fsspec are just `dict`s. It's like meta description of a file, not the file itself. The usual way in fsspect to read a file is this:

```python
fs.open('/remote/data/myfile.csv')
```

So, we can, I guess: read a file meta info, then create a file-like object with `fs.open()` and pass it into the `read_csv()` function, but I'm not sure if it makes sense.

Reply:
@rudolfix, ah, actually, no, it'll not work. `fsspec.open()` returns not a standard file-like object, so `duckdb` can't read it:

![Untitled2](https://github.com/dlt-hub/verified-sources/assets/26030071/703095d6-6d86-4fc5-a5ba-1a1926be1499)

Reply:
@rudolfix, a quite cumbersome piece of code may make it work:

```python
connection = duckdb.connect()

for file in files:
    if file["modification_date"] <= start_from:
        continue

    with fsspec.open(file["file_url"], mode="rb") as f:
        file_res = connection.read_csv(f)
        results += add_columns(file_res.columns, file_res.fetchall())
```